lint/checkstyle.bzl

```diff
@@ -0,0 +1,149 @@
+
+    ctx.actions.run_shell(
+        inputs = inputs,
+        outputs = outputs,
+        command = command.format(CHECKSTYLE = executable.path, stdout = stdout.path),
+        arguments = [args],
+        mnemonic = _MNEMONIC,
+        tools = [executable],
+        progress_message = "Linting %{label} with Checkstyle",
+    )
+
+# buildifier: disable=function-docstring
+def _checkstyle_aspect_impl(target, ctx):
+    if not should_visit(ctx.rule, ctx.attr._rule_kinds):
+        return []
+
+    files_to_lint = filter_srcs(ctx.rule)
+    outputs, info = output_files(_MNEMONIC, target, ctx)
+    if len(files_to_lint) == 0:
+        noop_lint_action(ctx, outputs)
+        return [info]
+
+    format_options = []
+    checkstyle_action(ctx, ctx.executable._checkstyle, files_to_lint, ctx.file._config, ctx.files._data, outputs.human.out, outputs.human.exit_code, format_options)
+    checkstyle_action(ctx, ctx.executable._checkstyle, files_to_lint, ctx.file._config, ctx.files._data, outputs.machine.out, outputs.machine.exit_code)
+    return [info]
+
+def lint_checkstyle_aspect(binary, config, data = [], rule_kinds = ["java_binary", "java_library"]):
+    """A factory function to create a linter aspect.
+
+    Attrs:
+        binary: a Checkstyle executable. Can be obtained from rules_java like so:
+
+            ```
+            java_binary(
+                name = "checkstyle",
+                main_class = "com.puppycrawl.tools.checkstyle.Main",
+                # Point to wherever you have the java_import rule defined, see our example
+                runtime_deps = ["@com_puppycrawl_tools_checkstyle"],
+            )
+            ```
+
+        config: the Checkstyle XML file
+    """
+    return aspect(
+        implementation = _checkstyle_aspect_impl,
+        # Edges we need to walk up the graph from the selected targets.
+        # Needed for linters that need semantic information like transitive type declarations.
+        # attr_aspects = ["deps"],
+        attrs = {
+            "_options": attr.label(
+                default = "//lint:options",
+                providers = [LintOptionsInfo],
+            ),
+            "_checkstyle": attr.label(
+                default = binary,
+                executable = True,
+                cfg = "exec",
+            ),
+            "_config": attr.label(
+                allow_single_file = True,
+                mandatory = True,
+                doc = "Config file",
+                default = config,
+            ),
+            "_data": attr.label_list(
+                doc = "Additional files to make available to Checkstyle such as any included XML files",
+                allow_files = True,
+                default = data,
+            ),
+            "_rule_kinds": attr.string_list(
+                default = rule_kinds,
+            ),
+        },
+    )
+
+def fetch_checkstyle():
+    http_jar(
+        name = "com_puppycrawl_tools_checkstyle",
```

Comment:
nit: is "puppycrawl" a vestigial name for the org that produced the tool originally? It seems like an odd name here.

Reply:
Yes, its the org name in the maven coordinates: https://mvnrepository.com/artifact/com.puppycrawl.tools/checkstyle

Reply:
I was following the naming of `fetch_pmd()` which uses `net_sourceforge_pmd`